lib/datasets/mnist.rb

```diff
@@ -45,9 +46,26 @@ def each(&block)
       image_path = cache_dir_path + target_file(:image)
       label_path = cache_dir_path + target_file(:label)
       base_url = self.class::BASE_URL
-
-      download(image_path, base_url + target_file(:image))
-      download(label_path, base_url + target_file(:label))
+      fallback_url = self.class::FALLBACK_URL
+
+      begin
+        download(image_path, base_url + target_file(:image))
+      rescue Net::HTTPClientException => error
+        if error.response.code == "403"
+          download(image_path, fallback_url + target_file(:image))
+        else
+          raise error
+        end
+      end
```

Comment:
Can we do this in `download`?

```ruby
      download(image_path,
               *base_urls.collect {|base_url| base_url + target_file(:image)})
      download(label_path,
               *base_urls.collect {|base_url| base_url + target_file(:label)})
```

Reply:
Thanks for your suggestion! How about this?:

```diff
diff --git a/lib/datasets/dataset.rb b/lib/datasets/dataset.rb
index 20b5a4e..d7fdc00 100644
--- a/lib/datasets/dataset.rb
+++ b/lib/datasets/dataset.rb
@@ -33,9 +33,13 @@ module Datasets
       @cache_path ||= CachePath.new(@metadata.id)
     end

-    def download(output_path, url, &block)
-      downloader = Downloader.new(url)
-      downloader.download(output_path, &block)
+    def download(output_path, *urls, &block)
+      urls.each do |url|
+        downloader = Downloader.new(url)
+        downloader.download(output_path, &block)
+      rescue Net::HTTPClientException => error
+        pp error # I didn't know what to do here...
+      end
     end

     def extract_bz2(bz2)
diff --git a/lib/datasets/mnist.rb b/lib/datasets/mnist.rb
index c4918e9..90a2cf2 100644
--- a/lib/datasets/mnist.rb
+++ b/lib/datasets/mnist.rb
@@ -42,27 +42,11 @@ module Datasets

       image_path = cache_dir_path + target_file(:image)
       label_path = cache_dir_path + target_file(:label)
-      base_url = self.class::BASE_URL
-      fallback_url = self.class::FALLBACK_URL
-
-      begin
-        download(image_path, base_url + target_file(:image))
-      rescue Net::HTTPClientException => error
-        if error.response.code == "403"
-          download(image_path, fallback_url + target_file(:image))
-        else
-          raise error
-        end
-      end
-      begin
-        download(label_path, base_url + target_file(:label))
-      rescue Net::HTTPClientException => error
-        if error.response.code == "403"
-          download(label_path, fallback_url + target_file(:label))
-        else
-          raise error
-        end
-      end
+
+      download(image_path,
+               *base_urls.collect { |base_url| base_url + target_file(:image) })
+      download(label_path,
+               *base_urls.collect { |base_url| base_url + target_file(:label) })

       open_data(image_path, label_path, &block)
     end
```

Reply:
I implemented in [784a1d1](https://github.com/red-data-tools/red-datasets/pull/196/commits/784a1d14abba507bcc9a161085562bc02fa56352). Could you check this change?